### PR TITLE
Compiler regression tests

### PR DESCRIPTION
**Pull request by hiiamboris**
These are the missing parts of PR #2191:
1. R/S compiler regression tests
2. Red compiler regression tests

I was hesitant for a long time to include these into the master branch. But after it caught 10 regressions at once I realized the impact it has.

Part 2 is only included into the `tests/run-all.r` (run only by the CI/CD setup), for it's pretty slow (about 5 minutes). It is split into five parts for the same reason. When your test №2000+ fails you don't want to start the whole run over. It's just more convenient to have it in parts in my experience.

Tests for this part should fail for now until we fix the december regression. Regressions are caused by https://github.com/red/red/commit/4414d6473cfea80aecf32c6b8aa5ee99b161715d : #2182 #1207 #1171 #1120 #1035 #1598 #837 #1974 #1935 #916

Once this PR is merged, we can close 2191 and make @rebolek happy :)

### CODE REVIEW

**Comment by greggirwin** [comment]
I'll let @dockimbel  or @qtxie make the call on this one. Given the results, I think it's good to include it, but we also need to look at testing speed overall, to see what we can do there.

**Comment by hiiamboris** [comment]
Yes. It's up to @dockimbel to first fix the stack thing that caused 10 tests to fail. Don't merge until that or no new automated builds :)

**Comment by dockimbel** [comment]
@hiiamboris Thanks for catching those regressions!

Indeed, we need to make the tests pass before including them in %run-all.r. As we run the whole test suite often dozens of times a day, the running speed is critical for our productivity. Keeping the running time as short as possible is high priority. Splitting the tests in several parts is a good option.

**Comment by hiiamboris** [comment]
Some of those old tickets were reopened. Isn't it enough?

**Comment by dockimbel** [comment]
@hiiamboris Given the number of regressions that all caused by the same commit, an umbrella "epic" ticket would be welcome.

**Comment by hiiamboris** [comment]
Interesting. Tests fail but CI/CD gives OK.

**Comment by dockimbel** [comment]
After the latest fixes, almost all the new regression tests are passing except for #3733 and #3773. See the following comments for details.

File: tests/source/compiler/regression-test-redc-5.r [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		--assert not crashed?
+		--assert script-error?
+
+		;; FIXME: this should not error out when compiled:
+		--compile-and-run-this-red {
+			do [f: func [/q] bind [1] context []]
+			f/q
+		}
+		--assert not crashed?
+		; --assert not script-error?
```
Comment (dockimbel) [inline tests/source/compiler/regression-test-redc-5.r]: This test is not a regression, it's a new issue and needs its own ticket.

File: tests/source/compiler/regression-test-redc-5.r [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		--assert compiler-error?
+
+		;; :mc = func [x][x], so `mc :mc` executing `x` applies it to an empty arg list => error
+		--compile-and-run-this-red {
+			#macro mc: func [x] [x]
+			probe quote (mc :mc)
+		}
+		--assert compiler-error?
+
```
Comment (dockimbel) [inline tests/source/compiler/regression-test-redc-5.r]: Those tests are invalid, compiled macros are run by **Rebol2**, not by Red. ;-)

Comment (hiiamboris) [reply tests/source/compiler/regression-test-redc-5.r]: Haha :) Right! I'll remove it then

File: tests/source/compiler/regression-test-redc-5.r (hunk 1 above)
Comment (hiiamboris) [reply tests/source/compiler/regression-test-redc-5.r]: Okay: https://github.com/red/red/issues/3945

Comment (hiiamboris) [reply tests/source/compiler/regression-test-redc-5.r]: It should pass the `not crashed?` test. The `not script-error?` test should be added by those who fix the new #3945 issue, to completely cover it.

**Comment by hiiamboris** [comment]
@dockimbel I've fixed the tests, however there's a recent (likely, July) R/S compiler regression that makes tests 273 and 275 of %regression-test-rsc.r fail on // operator:
```
*** Compilation Error: argument type mismatch on calling: //
*** expected: [number!], found: [float!]
```

**Comment by dockimbel** [comment]
@hiiamboris Support for `//` and `%` on floats has been removed, those operations are no more supported in R/S (see red/docs/red-system-specs-light.html#section-8.1). The reason is the lack of clear definition for such operations and difficulty to get consistent results on all platforms.

**Comment by dockimbel** [comment]
@hiiamboris Are we good now? Is there anything remaining to fix or change, or can we merge this PR now?

**Comment by hiiamboris** [comment]
@dockimbel yeah, all set.
The failure in on ARM only, unrelated to the PR.

**Pull request merged**